# Step 11. The same fit, for cohorts B and C

Step 02 fits cohort A. Steps 12–16 compare all three, so B and C need the identical fit:
same power, same `minModuleSize`, same merge height, same seed. Anything else and a difference
between cohorts could be a difference in parameters rather than in the data.

Reads `cohorts/`. Writes `data/run_artifacts/wgcna_B.rds` and `wgcna_C.rds`, in exactly the shape
step 02 writes `wgcna_A.rds`, so every later notebook can treat the three alike.

This is the slow step: two full `blockwiseModules` runs, roughly 80 seconds each.

Nothing here is tuned per cohort. The parameters come from steps 01–02 and are repeated verbatim:
the unsigned network, the power of 6, `minModuleSize = 20` and the merge threshold of 0.25 are the
source paper's published values, set out in step 02 in the paper's own words.

In [1]:
source("../src/paths.R")
suppressMessages(library(WGCNA))
options(stringsAsFactors = FALSE); enableWGCNAThreads(4)

POWER <- 6; MIN_MODULE <- 20; MERGE_CUT <- 0.25      # the published values, identical to step 02

# The curated panel and its control, resolved to SomaScan columns by step 01b.
pan  <- read.csv(art("ifn_panel.csv"), stringsAsFactors = FALSE)
IRG  <- pan$column[pan$set == "IRG-28"   & !is.na(pan$column)]
CTRL <- pan$column[pan$set == "NF-kB-11" & !is.na(pan$column)]

for (COHORT in c("B", "C")) {
  set.seed(SEED)
  X <- read.csv(coh("R_cohort-%s_log2_combat.csv", COHORT), row.names = 1, check.names = FALSE)
  m <- read.csv(coh("R_cohort-%s_meta.csv",        COHORT), row.names = 1, check.names = FALSE)
  m <- m[rownames(X), , drop = FALSE]

  net  <- blockwiseModules(X, power = POWER, networkType = "unsigned",
                           minModuleSize = MIN_MODULE, mergeCutHeight = MERGE_CUT,
                           numericLabels = TRUE, maxBlockSize = 8000, verbose = 0)
  mods <- labels2colors(net$colors)
  ME   <- moduleEigengenes(X, mods)$eigengenes

  # Name the interferon module the same way step 02b does for cohort A: the module
  # holding the most of the 28 interferon response genes, grey excluded because grey
  # is the unassigned pool and not a module. B and C are SEPARATE fits, so A's colour
  # cannot be carried over -- a WGCNA colour is only meaningful inside one fit.
  hit  <- mods[match(intersect(IRG, colnames(X)), colnames(X))]
  tab  <- sort(table(hit[hit != "grey"]), decreasing = TRUE)
  ifn  <- if (length(tab)) names(tab)[1] else NA_character_
  # The control: NF-kB-only genes should not be in it (de Jesus 2020).
  chit <- mods[match(intersect(CTRL, colnames(X)), colnames(X))]

  saveRDS(list(cohort = COHORT, X = as.matrix(X), meta = m, power = POWER,
               mods = mods, ME = ME, ifn = ifn, ifn_hits = tab, ctrl_hits = table(chit),
               minModuleSize = MIN_MODULE, mergeCutHeight = MERGE_CUT, seed = SEED),
          art("wgcna_%s.rds", COHORT))
  write.csv(data.frame(protein = colnames(X), module = mods),
            art("modules_%s.csv", COHORT), row.names = FALSE)

  cat(sprintf("cohort %s: n = %d, %d modules (excl. grey), %d reagents unassigned\n",
              COHORT, nrow(X), length(setdiff(unique(mods), "grey")), sum(mods == "grey")))
  cat(sprintf("  interferon module = %s, holding %s of %d IRG-28 reagents; %d in grey\n",
              ifn, if (length(tab)) tab[[1]] else 0L, length(intersect(IRG, colnames(X))),
              sum(hit == "grey")))
  cat(sprintf("  control: %d of %d NF-kB-only reagents in that module\n\n",
              sum(chit == ifn, na.rm = TRUE), length(intersect(CTRL, colnames(X)))))
}

Allowing parallel execution with up to 4 working processes.


cohort B: n = 87, 15 modules (excl. grey), 1658 reagents unassigned
  interferon module = turquoise, holding 5 of 12 IRG-28 reagents; 7 in grey
  control: 6 of 13 NF-kB-only reagents in that module

cohort C: n = 86, 37 modules (excl. grey), 1376 reagents unassigned
  interferon module = brown, holding 1 of 12 IRG-28 reagents; 7 in grey
  control: 1 of 13 NF-kB-only reagents in that module



**Explanation**
- `POWER`, `MIN_MODULE` and `MERGE_CUT` repeat step 02 verbatim. `SEED` now comes from
  `src/paths.R` rather than being written out again here, so one value governs the whole pipeline.
- `pan` is the panel step 01b wrote. `IRG` and `CTRL` are its two lists as SomaScan column names: the
  28 interferon response genes, and the 11 NF-kB-only control genes of de Jesus AA et al.
  *J Clin Invest* 2020;130:1669-1682.
- `blockwiseModules()` fits the network; `labels2colors()` turns WGCNA's numeric labels into its
  colour names; `moduleEigengenes()` returns the first principal component of each module.
- `hit` is the module of each curated reagent; `table()` counts them and `sort(decreasing = TRUE)`
  puts the largest first. `hit[hit != "grey"]` removes the unassigned pool before taking the largest,
  because `grey` is not a module. The count that went to grey is printed separately.
- `chit` does the same for the control. **The control is what makes the name mean something:** if
  NF-kB-only genes sat in the module too, it would be describing immune activation and not
  interferon.
- `tryCatch` and a minimum-hits threshold are gone. There was no citation for a threshold of three,
  so there is no threshold; the table and the control are printed and the reader judges them.
- The saved object now also carries `ifn_hits` and `ctrl_hits`, so a later notebook has the evidence
  and not only the answer.

**The over-representation test is not repeated here.** Step 02b runs
`ReactomePA::enrichPathway()` on cohort A, the method the source paper used. Repeating it for B's 15
and C's 37 modules is another 52 tests at about 15 seconds each, roughly 13 minutes, and it is not
run by default for that reason alone. The code is step 02b's, unchanged except for the cohort.

**Result. Under the published parameters the interferon module reproduces in cohort A, and in cohort
A only.**

| cohort | modules | grey | largest module | interferon module | its size | rank by size | IRG-28 in it | IRG-28 in grey | NF-kB controls in it |
|---|---|---|---|---|---|---|---|---|---|
| A | 33 | 1,107 | 2,847 (39%) | `skyblue` | 27 probes (23 proteins) | 28th | **6 of 12** | 1 | **0 of 13** |
| B | 15 | 1,658 | 4,522 (62%) | `turquoise` | **4,522 probes (4,168 proteins)** | **1st** | 5 of 12 | 7 | **6 of 13** |
| C | 37 | 1,376 | 3,409 (47%) | `brown` | 297 probes (290 proteins) | 3rd | **1 of 12** | 7 | 1 of 13 |

**Cohort A reproduces the published module closely.** `skyblue` holds 27 reagents against the paper's
35 proteins, and its hub by module membership is ISG15 `seq.14148.2` — the same SOMAmer the paper
names as the hub of its own interferon module, found here on a third of the samples by the same
parameters. No control reagent is in it. Step 02b also confirms it by over-representation
(Interferon Signaling, BH p = 0.00032).

**Cohort B has no interferon module.** Its answer is `turquoise`, which *is* its largest module:
4,522 reagents, 62% of the panel. It holds 5 of the 12 curated reagents while 7 went to grey, and it
contains **6 of the 13 NF-kB-only controls**. A module that holds half the controls discriminates
nothing. This is the same finding as before this notebook's parameters changed, and it is now starker.

**Cohort C has lost the interferon module it previously had.** The plurality is `brown` with **1 of
12** curated reagents, and 7 of the 12 went to grey. Under the signed network at power 12 with
`minModuleSize = 5`, cohort C gave a clean 15-reagent module holding 6 of 12 with no control in it.
At `minModuleSize = 20` that module is below the floor and its members are unassigned.

**What this says about the parameters, stated as a limitation rather than resolved here.** The
published `minModuleSize` of 20 was chosen on 207 samples. On 87 it sends 7 of the 12 curated
reagents to grey in two of three cohorts. Cohort A is the exception, not the rule. Batch composition
is near-identical across the three (69/18, 69/18, 69/17) and SLEDAI-2K medians are 2, 4 and 4, so this
is not a batch or disease-activity difference — it is which patients landed in which third of the
split. A floor of 20 cannot represent a module of 15, and at n ≈ 87 this module is sometimes smaller
than 20.

Two readings are available and the choice is not the notebook's to make: either `minModuleSize = 20`
is correct and cohorts B and C genuinely lack the module at this sample size, or 20 is too high for
n ≈ 87 and is suppressing a module that is present. The scan in step 02 is what distinguishes them,
and the paper's own sensitivity analysis went the other way, from 20 up to 30, never down.

## The three fits are not alike, and that is a result

Same study, same parameters, a random split of the same 260 donors:

| cohort | n | modules | unassigned | largest module |
|---|---|---|---|---|
| A | 87 | 33 | 1,107 | 2,847 (39.1%) |
| B | 87 | 15 | 1,658 | 4,522 (62.0%) |
| C | 86 | 37 | 1,376 | 3,409 (46.8%) |

B's network resolves into fewer than half as many modules as A's and puts **62% of the panel into a
single module**. That is not a parameter choice, the parameters are identical, and step 17 shows it is
not the patients either: batch composition is 69/18, 69/18, 69/17 by construction, and trait spread is
close across all three.

It is the reason cohort A yields 9 trait-associated modules where B yields 2 and C yields 1, and it is
the single most important caveat on everything in steps 12-18: module discovery at n ≈ 87 is not
stable across a random split of the same cohort.

## Where this was run

Rendered notebooks are committed, so each records the machine, the R and the
package versions that produced its output.


In [2]:
# Provenance. The rendered notebook is committed, so it records where it ran.
# The cell is identical in every notebook and is explained in step 00.
si <- Sys.info()
cat("run on      :", si[["nodename"]], "(", si[["sysname"]], si[["release"]], ")\n")
cat("date        :", format(Sys.time(), "%Y-%m-%d %H:%M %Z"), "\n")
cat("R           :", R.version.string, "|", R.version$platform, "\n")
cat("R comes from:", R.home(), "\n")
cat("packages    :\n")
for (p in c("WGCNA", "ComplexHeatmap", "circlize", "ReactomePA", "reactome.db",
            "org.Hs.eg.db", "sva", "VarSelLCM"))
  cat(sprintf("   %-16s %s\n", p,
              tryCatch(as.character(packageVersion(p)), error = function(e) "not installed")))

run on      : Annes-MacBook-Pro-193.local ( Darwin 27.0.0 )


date        : 2026-09-28 01:02 EDT 


R           : R version 4.4.3 (2025-02-28) | x86_64-apple-darwin13.4.0 


R comes from: /Users/adeslatt/miniforge3/envs/endotypes-proteomics/lib/R 


packages    :


   WGCNA            1.74
   ComplexHeatmap   2.22.0
   circlize         0.4.18
   ReactomePA       1.50.0
   reactome.db      1.89.0
   org.Hs.eg.db     3.20.0
   sva              3.54.0
   VarSelLCM        2.1.3.2
